halo2-base/src/gates/flex_gate.rs

```diff
@@ -709,6 +709,25 @@ pub trait GateInstructions<F: ScalarField> {
         self.mul(ctx, a, b)
     }
 
+    fn and_many(
+        &self,
+        ctx: &mut Context<F>,
+        values: impl IntoIterator<Item = QuantumCell<F>>,
+    ) -> AssignedValue<F> {
+        let values = values.into_iter();
+        let (len, hi) = values.size_hint();
+        debug_assert!(len > 2, "at least 2 elements to perform AND operation");
+        debug_assert_eq!(Some(len), hi);
+
+        let mut a = Value::known(F::one());
+        let cells = values.flat_map(|b| {
+            let prev_res = a.clone();
+            a = a.zip(b.value()).map(|(a, b)| a * b);
+            [Constant(F::zero()), Witness(prev_res), b, Witness(a)]
+        });
+        self.assign_region_last(ctx, cells, (0..len).map(|i| (4 * i as isize, None)))
```

Comment:
1. It seems you did not add equality offsets `[(3, 5), (7, 9), ....]`?
2. You use `values.len() * 4` cells here, why do you just call `.add()` multiple times which only need `(values().len() - 1)*4` cells. 

    - The cells in your impl looks like `[0, 1, a, a, 0, a, b, a*b, 0, a*b, c, a*b*c, ...]`. 

   - Using `add()` in a loop looks like `[0, a, b, a*b, 0, a*b, c, a*b*c, ....]`.

Reply:
I can use the `mul()` method multiple times, is that what you meant?